keps/sig-cloud-provider/2025-extend-konnectivity-for-both-directions/README.md

```diff
@@ -0,0 +1,678 @@
+and make progress.
+-->
+* Define a mechanism for exchanging authentication information used for establishing the secure channels between agents and server (e.g. certificates, tokens). 
+* Define a solution involving less than one agent per node. 
+
+## Definitions
+
+* **Master Network** An IP reachable network space which contains the master components, such as Kubernetes API Server, Connectivity Proxy and ETCD server.
+* **Node Network** An IP reachable network space which contains all the clusters Nodes, for alpha. Worth noting that the Node Network may be completely disjoint from the Master network. It may have overlapping IP addresses to the Master Network or other means of network isolation. Direct IP routability between cluster and master networks should not be assumed. Later versions may relax the all node requirement to some.
+* **KAS** Kubernetes API Server, responsible for serving the Kubernetes API to clients.
+* **Konnectivity Server** The proxy server which runs in the master network. It has a secure channel established to the cluster network. It could work on either a HTTP Connect mechanism or gRPC. If the former it would expose a gRPC interface to KAS to provide connectivity service. If the latter it would use standard HTTP Connect. Formerly known as the Network Proxy Server.
+* **Konnectivity Agent** A proxy agent which runs in the node network for establishing the tunnel. Formerly known as the Network Proxy Agent.
+* **Flat Network** A network where traffic can be successfully routed using IP. Implies no overlapping (i.e. shared) IPs on the network.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+Currently the Konnectivity Server is accepting requests from the KAS either with the gRPC or the HTTP Connect interfaces and is taking care of forwarding the traffic to the Konnectivity Agent using the previously established connections (initiated by the Agents). 
+
+In order to enable traffic from Kubelets and Pods running on Master Network, the Konnectivity Agents have to expose an endpoint that will be listening on a specific port for each of the destinations on the Master Network. As opposed to the traffic flowing from the Master Network to the Node Network, the Konnectivity Agent should act transparently: From a Kubelets or Pods standpoint, the Konnectivity Agent should be the final destination instead of acting as a proxy. 
+
+The reason why we do not intend to use the same strategy used in the other direction is that we do not have control over the clients using the Kubernetes default service to communicate with the KAS.
+
```

Comment:
I think we also need to detail a few key points about the konnectivity server. 
1) The feature needs to be enable on the server (and agent?) for it to be enabled.
2) The konnectivity server will not forward traffic destined for anything other than the KAS. (So we will not allow traffic to be sent to ETCD via this mechanism).

Reply:
@cheftako re:

>2. The konnectivity server will not forward traffic destined for anything other than the KAS. (So we will not allow traffic to be sent to ETCD via this mechanism).

I'm a bit confused here: the previous paragraph describes how port pairs can be defined to route traffic from the node to the master network for what seem to be arbitrary (though well-controlled) destinations on the master network/host. However, your comment makes me believe you propose traffic should be forwarded to the KAS only. What am I missing?

One use case I have in mind here is Cilium: it can be configured to use the etcd cluster co-located on the master for state-keeping purposes. I was hoping that Konnectivity could allow for this (since the alternative of directly exposing etcd to the public Internet would be rather undesirable).

Reply:
Probably the misunderstanding comes from the fact that our original plan was to allow reaching arbitrary destinations, but we did not have a real use-case for providing this flexibility (https://github.com/kubernetes/enhancements/pull/2025#discussion_r497760884).

We have two options:
1. Either we start with KAS support only, and we extend it in the future in case of need (in this case we have to update the paragraph describing the agent flags as @timoreimann pointed out to make it clearer).
2. We allow reaching arbitrary destinations in the master cluster (only if explicitly allowed).
Personally, I'm more in favor of 2, but I don't have a strong opinion.

Reply:
For me the ability to reach the KAS from the node network is definitely priority no. 1. If punting on the idea to support arbitrary destinations helps us getting the KEP approved more quickly and increasing our chances to get something workable into the 1.21 release (which the milestone is currently set to), I'm totally okay with keeping the scope narrow in the first iteration.

Reply:
I would suggest the konnectivity server be configured for what traffic it will allow to be placed on its local network. I would suggest the minimum options be None (the default) and KAS only. It you want more options than that, that's fine but I believe those two would be the minimum.

Reply:
yeah, we addressed that with the `allowed-destination` flag, if not present no traffic will be allowed.